File: app/src/main/java/com/xeniac/warrantyroster/mainactivity/setingsfragment/SettingsFragment.java
package com.xeniac.warrantyroster.mainactivity.setingsfragment;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.airbnb.lottie.LottieDrawable;
import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.request.RequestHeaders;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.xeniac.warrantyroster.Constants;
import com.xeniac.warrantyroster.NetworkHelper;
import com.xeniac.warrantyroster.R;
import com.xeniac.warrantyroster.databinding.FragmentSettingsBinding;
import com.xeniac.warrantyroster.landingactivity.LandingActivity;
import com.xeniac.warrantyroster.mainactivity.MainActivity;

import ir.tapsell.plus.AdHolder;
import ir.tapsell.plus.AdRequestCallback;
import ir.tapsell.plus.AdShowListener;
import ir.tapsell.plus.TapsellPlus;
import ir.tapsell.plus.TapsellPlusInitListener;
import ir.tapsell.plus.model.AdNetworkError;
import ir.tapsell.plus.model.AdNetworks;
import ir.tapsell.plus.model.TapsellPlusAdModel;

public class SettingsFragment extends Fragment {

    private FragmentSettingsBinding settingsBinding;
    private View view;
    private Activity activity;
    private Context context;
    private NavController navController;

    private SharedPreferences settingsPrefs;
    private String currentLanguage;
    private String currentCountry;
    private int currentTheme;
    private String accountEmail;
    private ColorStateList verifyEmailBackgroundTint;

    private int requestAdCounter;
    private String responseId;

    public SettingsFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        settingsBinding = FragmentSettingsBinding.inflate(inflater, container, false);
        view = settingsBinding.getRoot();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        destroyAd();
        settingsBinding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        activity = getActivity();
        context = getContext();
        navController = Navigation.findNavController(view);
        ((MainActivity) context).showNavBar();

        settingsPrefs = context.getSharedPreferences(Constants.PREFERENCE_SETTINGS, Context.MODE_PRIVATE);
        currentLanguage = settingsPrefs.getString(Constants.PREFERENCE_LANGUAGE_KEY, "en");
        currentCountry = settingsPrefs.getString(Constants.PREFERENCE_COUNTRY_KEY, "US");
        currentTheme = settingsPrefs.getInt(Constants.PREFERENCE_THEME_KEY, 0);

        getAccountDetails();
        setCurrentLanguageText();
        setCurrentThemeText();
        verifyOnClick();
        changeEmailOnClick();
        changePasswordOnClick();
        languageOnClick();
        themeOnClick();
        privacyPolicyOnClick();
        logoutOnClick();
        adInit();
    }

    private void getAccountDetails() {
        if (NetworkHelper.hasNetworkAccess(context)) {
            getAccountDetailsQuery();
        } else {
            hideAccountDetails();
            hideLoadingAnimation();
            Snackbar.make(view, context.getResources().getString(R.string.network_error_connection),
                    BaseTransientBottomBar.LENGTH_INDEFINITE)
                    .setAction(context.getResources().getString(R.string.network_error_retry), v -> getAccountDetails()).show();
        }
    }

    private void getAccountDetailsQuery() {
        showLoadingAnimation();

        SharedPreferences loginPrefs = context.getSharedPreferences(Constants.PREFERENCE_LOGIN, Context.MODE_PRIVATE);
        String userToken = loginPrefs.getString(Constants.PREFERENCE_USER_TOKEN_KEY, null);

        ApolloClient apolloClient = ApolloClient.builder()
                .serverUrl(Constants.URL_GRAPHQL)
                .build();

        apolloClient.query(new GetAccountDetailsQuery())
                .toBuilder().requestHeaders(RequestHeaders.builder().addHeader("Authorization", "Bearer " + userToken).build())
                .build()
                .enqueue(new ApolloCall.Callback<GetAccountDetailsQuery.Data>() {
                    @Override
                    public void onResponse(@NonNull Response<GetAccountDetailsQuery.Data> response) {
                        if (settingsBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();

                                if (!response.hasErrors()) {
                                    Log.i("getAccountDetails", "onResponse: " + response);
                                    accountEmail = response.getData().getAccountDetails().email();
                                    setAccountDetails(accountEmail,
                                            response.getData().getAccountDetails().is_email_verified());
                                } else {
                                    Log.e("getAccountDetails", "onResponse Errors: " + response.getErrors());
                                    Snackbar.make(view, context.getResources().getString(R.string.network_error_response),
                                            BaseTransientBottomBar.LENGTH_INDEFINITE)
                                            .setAction(context.getResources().getString(R.string.network_error_retry), v -> getAccountDetails()).show();
                                }
                            });
                        }
                    }

                    @Override
                    public void onFailure(@NonNull ApolloException e) {
                        Log.e("getAccountDetails", "onFailure: " + e.getMessage());
                        if (settingsBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();
                                hideAccountDetails();
                                Snackbar.make(view, context.getResources().getString(R.string.network_error_failure),
                                        BaseTransientBottomBar.LENGTH_LONG).show();
                            });
                        }
                    }
                });
    }

    private void setAccountDetails(String email, boolean isEmailVerified) {
        showAccountDetails();
        settingsBinding.tvSettingsAccountEmail.setText(email);

        if (isEmailVerified) {
            settingsBinding.btnSettingsAccountEmail.setClickable(false);
            settingsBinding.btnSettingsAccountEmail.setText(context.getResources().getString(R.string.settings_btn_account_verified));
            settingsBinding.btnSettingsAccountEmail.setTextColor(context.getResources().getColor(R.color.green));
            settingsBinding.btnSettingsAccountEmail.setBackgroundTintList(ColorStateList.valueOf(context.getResources().getColor(R.color.green20)));
            settingsBinding.ivSettingsAccountEmail.setBackgroundColor(context.getResources().getColor(R.color.green10));
            settingsBinding.lavSettingsAccountEmail.setRepeatCount(0);
            settingsBinding.lavSettingsAccountEmail.setSpeed(0.60f);
            settingsBinding.lavSettingsAccountEmail.setAnimation(R.raw.anim_account_verified);
        } else {
            settingsBinding.btnSettingsAccountEmail.setClickable(true);
            settingsBinding.btnSettingsAccountEmail.setText(context.getResources().getString(R.string.settings_btn_account_verify));
            settingsBinding.btnSettingsAccountEmail.setTextColor(context.getResources().getColor(R.color.blue));
            settingsBinding.btnSettingsAccountEmail.setBackgroundTintList(ColorStateList.valueOf(context.getResources().getColor(R.color.blue20)));
            settingsBinding.ivSettingsAccountEmail.setBackgroundColor(context.getResources().getColor(R.color.red10));
            settingsBinding.lavSettingsAccountEmail.setRepeatCount(LottieDrawable.INFINITE);
            settingsBinding.lavSettingsAccountEmail.setSpeed(1.00f);
            settingsBinding.lavSettingsAccountEmail.setAnimation(R.raw.anim_account_not_verified);
        }

        settingsBinding.lavSettingsAccountEmail.playAnimation();
        verifyEmailBackgroundTint = settingsBinding.btnSettingsAccountEmail.getBackgroundTintList();
    }

    private void setCurrentLanguageText() {
        switch (currentLanguage) {
            case "en":
                settingsBinding.tvSettingsSettingsLanguageCurrent.setText(
                        context.getResources().getString(R.string.settings_text_settings_language_english));
                break;
            case "fa":
                settingsBinding.tvSettingsSettingsLanguageCurrent.setText(
                        context.getResources().getString(R.string.settings_text_settings_language_persian));
                break;
        }
    }

    private void setCurrentThemeText() {
        switch (currentTheme) {
            case 0:
                settingsBinding.tvSettingsSettingsThemeCurrent.setText(
                        context.getResources().getString(R.string.settings_text_settings_theme_default));
                break;
            case 1:
                settingsBinding.tvSettingsSettingsThemeCurrent.setText(
                        context.getResources().getString(R.string.settings_text_settings_theme_light));
                break;
            case 2:
                settingsBinding.tvSettingsSettingsThemeCurrent.setText(
                        context.getResources().getString(R.string.settings_text_settings_theme_dark));
                break;
        }
    }

    private void verifyOnClick() {
        settingsBinding.btnSettingsAccountEmail.setOnClickListener(view ->
                sendVerificationEmail());
    }

    private void sendVerificationEmail() {
        if (NetworkHelper.hasNetworkAccess(context)) {
            sendVerificationEmailMutation();
        } else {
            hideVerificationLoadingAnimation();
            Snackbar.make(view, context.getResources().getString(R.string.network_error_connection),
                    BaseTransientBottomBar.LENGTH_INDEFINITE)
                    .setAction(context.getResources().getString(R.string.network_error_retry), v -> sendVerificationEmail()).show();
        }
    }

    private void sendVerificationEmailMutation() {
        showVerificationLoadingAnimation();

        SharedPreferences loginPrefs = context.getSharedPreferences(Constants.PREFERENCE_LOGIN, Context.MODE_PRIVATE);
        String userToken = loginPrefs.getString(Constants.PREFERENCE_USER_TOKEN_KEY, null);

        ApolloClient apolloClient = ApolloClient.builder()
                .serverUrl(Constants.URL_GRAPHQL)
                .build();

        apolloClient.mutate(new SendVerificationEmailMutation())
                .toBuilder().requestHeaders(RequestHeaders.builder().addHeader("Authorization", "Bearer " + userToken).build())
                .build()
                .enqueue(new ApolloCall.Callback<SendVerificationEmailMutation.Data>() {
                    @Override
                    public void onResponse(@NonNull Response<SendVerificationEmailMutation.Data> response) {
                        if (settingsBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideVerificationLoadingAnimation();

                                if (!response.hasErrors()) {
                                    Log.i("sendVerificationEmail", "onResponse: " + response);
                                    MaterialAlertDialogBuilder sendVerificationEmailDialogBuilder = new MaterialAlertDialogBuilder(context)
                                            .setMessage(context.getResources().getString(R.string.settings_dialog_message))
                                            .setPositiveButton(context.getResources().getString(R.string.settings_dialog_positive), (dialog, which) -> {
                                            });
                                    sendVerificationEmailDialogBuilder.show();
                                } else {
                                    Log.e("sendVerificationEmail", "onResponse Errors: " + response.getErrors());
                                    Snackbar.make(view, context.getResources().getString(R.string.network_error_response),
                                            BaseTransientBottomBar.LENGTH_INDEFINITE)
                                            .setAction(context.getResources().getString(R.string.network_error_retry), v -> sendVerificationEmail())
                                            .show();
                                }
                            });
                        }
                    }

                    @Override
                    public void onFailure(@NonNull ApolloException e) {
                        Log.e("sendVerificationEmail", "onFailure: " + e.getMessage());
                        if (settingsBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideVerificationLoadingAnimation();
                                Snackbar.make(view, context.getResources().getString(R.string.network_error_failure),
                                        BaseTransientBottomBar.LENGTH_LONG).show();
                            });
                        }
                    }
                });
    }

    private void changeEmailOnClick() {
        settingsBinding.clSettingsAccountChangeEmail.setOnClickListener(view -> {
            SettingsFragmentDirections.ActionSettingsFragmentToChangeEmailFragment action =
                    SettingsFragmentDirections.actionSettingsFragmentToChangeEmailFragment(accountEmail);
            navController.navigate(action);
        });
    }

    private void changePasswordOnClick() {
        settingsBinding.clSettingsAccountChangePassword.setOnClickListener(view ->
                navController.navigate(R.id.action_settingsFragment_to_changePasswordFragment));
    }

    private void languageOnClick() {
        settingsBinding.clSettingsSettingsLanguage.setOnClickListener(view ->
                Toast.makeText(context, "Language", Toast.LENGTH_SHORT).show());
    }

    private void themeOnClick() {
        settingsBinding.clSettingsSettingsTheme.setOnClickListener(view -> {

            String[] themeItems = new String[]{context.getResources().getString(R.string.settings_text_settings_theme_default),
                    context.getResources().getString(R.string.settings_text_settings_theme_light),
                    context.getResources().getString(R.string.settings_text_settings_theme_dark)};

            MaterialAlertDialogBuilder themeDialogBuilder = new MaterialAlertDialogBuilder(context)
                    .setTitle(context.getResources().getString(R.string.settings_text_settings_theme))
                    .setSingleChoiceItems(themeItems, currentTheme, (dialogInterface, i) -> {
                        switch (i) {
                            case 0:
                                if (AppCompatDelegate.getDefaultNightMode() != AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM) {
                                    settingsPrefs.edit().putInt(Constants.PREFERENCE_THEME_KEY, i).apply();
                                    AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM);
                                }
                                break;
                            case 1:
                                if (AppCompatDelegate.getDefaultNightMode() != AppCompatDelegate.MODE_NIGHT_NO) {
                                    settingsPrefs.edit().putInt(Constants.PREFERENCE_THEME_KEY, i).apply();
                                    AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                                }
                                break;
                            case 2:
                                if (AppCompatDelegate.getDefaultNightMode() != AppCompatDelegate.MODE_NIGHT_YES) {
                                    settingsPrefs.edit().putInt(Constants.PREFERENCE_THEME_KEY, i).apply();
                                    AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                                }
                                break;
                        }
                        currentTheme = i;
                        setCurrentThemeText();
                        dialogInterface.dismiss();
                    });
            themeDialogBuilder.show();
        });
    }

    private void privacyPolicyOnClick() {
        settingsBinding.clSettingsSettingsPrivacyPolicy.setOnClickListener(view ->
                Toast.makeText(context, "Privacy Policy", Toast.LENGTH_SHORT).show());
    }

    private void logoutOnClick() {
        settingsBinding.btnSettingsLogout.setOnClickListener(view -> {
            SharedPreferences.Editor editor = context
                    .getSharedPreferences(Constants.PREFERENCE_LOGIN, Context.MODE_PRIVATE).edit();
            editor.remove(Constants.PREFERENCE_USER_TOKEN_KEY).apply();
            editor.remove(Constants.PREFERENCE_IS_LOGGED_IN_KEY).apply();

            startActivity(new Intent(context, LandingActivity.class));
            activity.finish();
        });
    }

    private void showAccountDetails() {
        settingsBinding.groupSettingsAccount.setVisibility(View.VISIBLE);
    }

    private void hideAccountDetails() {
        settingsBinding.groupSettingsAccount.setVisibility(View.GONE);
    }

    private void showLoadingAnimation() {
        settingsBinding.lpiSettingsAccountEmail.setVisibility(View.VISIBLE);

    }

    private void hideLoadingAnimation() {
        settingsBinding.lpiSettingsAccountEmail.setVisibility(View.GONE);
    }

    private void showVerificationLoadingAnimation() {
        settingsBinding.btnSettingsAccountEmail.setClickable(false);
        settingsBinding.btnSettingsAccountEmail.setText(null);
        settingsBinding.btnSettingsAccountEmail.setBackgroundTintList(null);
        settingsBinding.cpiSettingsAccountEmailVerification.setVisibility(View.VISIBLE);

    }

    private void hideVerificationLoadingAnimation() {
        settingsBinding.cpiSettingsAccountEmailVerification.setVisibility(View.GONE);
        settingsBinding.btnSettingsAccountEmail.setBackgroundTintList(verifyEmailBackgroundTint);
        settingsBinding.btnSettingsAccountEmail.setText(context.getResources().getString(R.string.settings_btn_account_verify));
        settingsBinding.btnSettingsAccountEmail.setClickable(true);
    }

    private void adInit() {
        TapsellPlus.initialize(context, Constants.TAPSELL_KEY, new TapsellPlusInitListener() {
            @Override
            public void onInitializeSuccess(AdNetworks adNetworks) {
                Log.i("adInit", "onInitializeSuccess: " + adNetworks.name());
                AdHolder adHolder = TapsellPlus.createAdHolder(activity,
                        settingsBinding.flSettingsAdContainer, R.layout.ad_banner_settings);
                requestAdCounter = 0;
                requestNativeAd(adHolder);
            }

            @Override
            public void onInitializeFailed(AdNetworks adNetworks, AdNetworkError adNetworkError) {
                Log.e("adInit", "onInitializeFailed: " + adNetworks.name() + ", error: " + adNetworkError.getErrorMessage());
            }
        });
    }

    private void requestNativeAd(AdHolder adHolder) {
        TapsellPlus.requestNativeAd(activity, Constants.SETTINGS_NATIVE_ZONE_ID,
                new AdRequestCallback() {
                    @Override
                    public void response(TapsellPlusAdModel tapsellPlusAdModel) {
                        super.response(tapsellPlusAdModel);
                        Log.i("requestNativeAd", "response: " + tapsellPlusAdModel.toString());
                        responseId = tapsellPlusAdModel.getResponseId();
                        showNativeAd(adHolder, responseId);
                    }

                    @Override
                    public void error(String s) {
                        super.error(s);
                        Log.e("requestNativeAd", "error: " + s);
                        if (requestAdCounter < 3) {
                            requestAdCounter++;
                            requestNativeAd(adHolder);
                        }
                    }
                });
    }

    private void showNativeAd(AdHolder adHolder, String responseId) {
        settingsBinding.groupSettingsAd.setVisibility(View.VISIBLE);

        TapsellPlus.showNativeAd(activity, responseId, adHolder,
                new AdShowListener() {
                    @Override
                    public void onOpened(TapsellPlusAdModel tapsellPlusAdModel) {
                        super.onOpened(tapsellPlusAdModel);
                    }

                    @Override
                    public void onClosed(TapsellPlusAdModel tapsellPlusAdModel) {
                        super.onClosed(tapsellPlusAdModel);
                    }
                });
    }

    private void destroyAd() {
        if (responseId != null) {
            TapsellPlus.destroyNativeBanner(activity, responseId);
        }
    }
}

File: app/src/main/java/com/xeniac/warrantyroster_manager/mainactivity/settingsfragment/SettingsFragment.kt
package com.xeniac.warrantyroster_manager.mainactivity.settingsfragment

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.Toast
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.navigation.NavController
import androidx.navigation.Navigation
import com.airbnb.lottie.LottieDrawable
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_INDEFINITE
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_LONG
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.xeniac.warrantyroster_manager.util.NetworkHelper
import com.xeniac.warrantyroster_manager.R
import com.xeniac.warrantyroster_manager.databinding.FragmentSettingsBinding
import com.xeniac.warrantyroster_manager.landingactivity.LandingActivity
import com.xeniac.warrantyroster_manager.mainactivity.MainActivity
import com.xeniac.warrantyroster_manager.util.Constants.Companion.PREFERENCE_COUNTRY_KEY
import com.xeniac.warrantyroster_manager.util.Constants.Companion.PREFERENCE_IS_LOGGED_IN_KEY
import com.xeniac.warrantyroster_manager.util.Constants.Companion.PREFERENCE_LANGUAGE_KEY
import com.xeniac.warrantyroster_manager.util.Constants.Companion.PREFERENCE_LOGIN
import com.xeniac.warrantyroster_manager.util.Constants.Companion.PREFERENCE_SETTINGS
import com.xeniac.warrantyroster_manager.util.Constants.Companion.PREFERENCE_THEME_KEY
import com.xeniac.warrantyroster_manager.util.Constants.Companion.SETTINGS_NATIVE_ZONE_ID
import com.xeniac.warrantyroster_manager.util.Constants.Companion.TAPSELL_KEY
import com.xeniac.warrantyroster_manager.util.Constants.Companion.URL_PRIVACY_POLICY
import ir.tapsell.plus.*
import ir.tapsell.plus.model.AdNetworkError
import ir.tapsell.plus.model.AdNetworks
import ir.tapsell.plus.model.TapsellPlusAdModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext

class SettingsFragment : Fragment(R.layout.fragment_settings) {

    private var _binding: FragmentSettingsBinding? = null
    private val binding get() = _binding!!
    private lateinit var navController: NavController

    private val firebaseAuth = FirebaseAuth.getInstance()
    private val currentUser = firebaseAuth.currentUser

    private lateinit var settingsPrefs: SharedPreferences
    private lateinit var currentLanguage: String
    private lateinit var currentCountry: String
    private var currentTheme: Int = 0

    private var requestAdCounter = 0
    private var responseId: String? = null

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentSettingsBinding.bind(view)
        navController = Navigation.findNavController(view)
        (requireContext() as MainActivity).showNavBar()

        getCurrentSettings()
        getAccountDetails()
        setCurrentLanguageText()
        setCurrentThemeText()
        verifyOnClick()
        changeEmailOnClick()
        changePasswordOnClick()
        languageOnClick()
        themeOnClick()
        privacyPolicyOnClick()
        logoutOnClick()
        adInit()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        destroyAd()
        _binding = null
    }

    private fun getCurrentSettings() {
        settingsPrefs = requireContext()
            .getSharedPreferences(PREFERENCE_SETTINGS, Context.MODE_PRIVATE)
        currentLanguage = settingsPrefs
            .getString(PREFERENCE_LANGUAGE_KEY, "en").toString()
        currentCountry = settingsPrefs.getString(PREFERENCE_COUNTRY_KEY, "US").toString()
        currentTheme = settingsPrefs.getInt(PREFERENCE_THEME_KEY, 0)
    }

    private fun getAccountDetails() = CoroutineScope(Dispatchers.IO).launch {
        try {
            currentUser?.let {
                var email = it.email.toString()
                var isVerified = it.isEmailVerified
                Log.i("getAccountDetails", "Current user is $email and isVerified: $isVerified")

                withContext(Dispatchers.Main) {
                    setAccountDetails(email, isVerified)
                }

                if (NetworkHelper.hasNetworkAccess(requireContext())) {
                    it.reload().await()
                    if (email != it.email || isVerified != it.isEmailVerified) {
                        email = it.email.toString()
                        isVerified = it.isEmailVerified
                        Log.i(
                            "getAccountDetails",
                            "Updated current user is $$email and isVerified: $isVerified"
                        )

                        withContext(Dispatchers.Main) {
                            setAccountDetails(email, isVerified)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            Log.i("getAccountDetails", "Exception: ${e.message}")
        }
    }

    private fun setAccountDetails(email: String, isEmailVerified: Boolean) {
        binding.tvAccountEmail.text = email

        if (isEmailVerified) {
            binding.btnAccountVerification.isClickable = false
            binding.btnAccountVerification.text =
                requireContext().getString(R.string.settings_btn_account_verified)
            binding.btnAccountVerification.setTextColor(
                ContextCompat.getColor(requireContext(), R.color.green)
            )
            binding.btnAccountVerification.backgroundTintList =
                ContextCompat.getColorStateList(requireContext(), R.color.green20)
            binding.ivAccountEmail.setBackgroundColor(
                ContextCompat.getColor(requireContext(), R.color.green10)
            )
            binding.lavAccountVerification.repeatCount = 0
            binding.lavAccountVerification.speed = 0.60f
            binding.lavAccountVerification.setAnimation(R.raw.anim_account_verified)
        } else {
            binding.btnAccountVerification.isClickable = true
            binding.btnAccountVerification.text =
                requireContext().getString(R.string.settings_btn_account_verify)
            binding.btnAccountVerification.setTextColor(
                ContextCompat.getColor(requireContext(), R.color.blue)
            )
            binding.btnAccountVerification.backgroundTintList =
                ContextCompat.getColorStateList(requireContext(), R.color.blue20)
            binding.ivAccountEmail.setBackgroundColor(
                ContextCompat.getColor(requireContext(), R.color.red10)
            )
            binding.lavAccountVerification.repeatCount = LottieDrawable.INFINITE
            binding.lavAccountVerification.speed = 1.00f
            binding.lavAccountVerification.setAnimation(R.raw.anim_account_not_verified)
        }
        binding.lavAccountVerification.playAnimation()
    }

    private fun setCurrentLanguageText() {
        when (currentLanguage) {
            "en" -> {
                binding.tvSettingsLanguageCurrent.text =
                    requireContext().getString(R.string.settings_text_settings_language_english)
            }
            "fa" -> {
                binding.tvSettingsLanguageCurrent.text =
                    requireContext().getString(R.string.settings_text_settings_language_persian)
            }
        }
    }

    private fun setCurrentThemeText() {
        when (currentTheme) {
            0 -> {
                binding.tvSettingsThemeCurrent.text =
                    requireContext().getString(R.string.settings_text_settings_theme_default)
            }
            1 -> {
                binding.tvSettingsThemeCurrent.text =
                    requireContext().getString(R.string.settings_text_settings_theme_light)
            }
            2 -> {
                binding.tvSettingsThemeCurrent.text =
                    requireContext().getString(R.string.settings_text_settings_theme_dark)
            }
        }
    }

    private fun verifyOnClick() = binding.btnAccountVerification.setOnClickListener {
        sendVerificationEmail()
    }

    private fun sendVerificationEmail() {
        if (NetworkHelper.hasNetworkAccess(requireContext())) {
            sendVerificationEmailAuth()
        } else {
            hideLoadingAnimation()
            Snackbar.make(
                binding.root,
                requireContext().getString(R.string.network_error_connection),
                LENGTH_INDEFINITE
            ).apply {
                setAction(requireContext().getString(R.string.network_error_retry)) { sendVerificationEmail() }
                show()
            }
        }
    }

    private fun sendVerificationEmailAuth() {
        showLoadingAnimation()
        CoroutineScope(Dispatchers.IO).launch {
            try {
                currentUser?.let {
                    it.sendEmailVerification().await()
                    Log.i("sendVerificationEmail", "Email sent to ${it.email}")
                    withContext(Dispatchers.Main) {
                        hideLoadingAnimation()

                        MaterialAlertDialogBuilder(requireContext()).apply {
                            setMessage(requireContext().getString(R.string.settings_dialog_message))
                            setPositiveButton(requireContext().getString(R.string.settings_dialog_positive)) { _, _ -> }
                        }.show()
                    }
                }
            } catch (e: Exception) {
                Log.e("sendVerificationEmail", "Exception: ${e.message}")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    Snackbar.make(
                        binding.root,
                        requireContext().getString(R.string.network_error_failure),
                        LENGTH_LONG
                    ).show()
                }
            }
        }
    }

    private fun changeEmailOnClick() = binding.clAccountChangeEmail.setOnClickListener {
        navController.navigate(R.id.action_settingsFragment_to_changeEmailFragment)
    }

    private fun changePasswordOnClick() =
        binding.clAccountChangePassword.setOnClickListener {
            navController.navigate(R.id.action_settingsFragment_to_changePasswordFragment)
        }

    private fun languageOnClick() = binding.clSettingsLanguage.setOnClickListener {
        Toast.makeText(requireContext(), "Language", Toast.LENGTH_SHORT).show()
    }

    private fun themeOnClick() = binding.clSettingsTheme.setOnClickListener {
        val themeItems = arrayOf(
            requireContext().getString(R.string.settings_text_settings_theme_default),
            requireContext().getString(R.string.settings_text_settings_theme_light),
            requireContext().getString(R.string.settings_text_settings_theme_dark)
        )

        MaterialAlertDialogBuilder(requireContext()).apply {
            setTitle(requireContext().getString(R.string.settings_text_settings_theme))
            setSingleChoiceItems(themeItems, currentTheme) { dialogInterface, i ->
                when (i) {
                    0 -> {
                        if (AppCompatDelegate.getDefaultNightMode() != AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM) {
                            settingsPrefs.edit().putInt(PREFERENCE_THEME_KEY, i)
                                .apply()
                            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
                        }
                    }
                    1 -> {
                        if (AppCompatDelegate.getDefaultNightMode() != AppCompatDelegate.MODE_NIGHT_NO) {
                            settingsPrefs.edit().putInt(PREFERENCE_THEME_KEY, i)
                                .apply()
                            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
                        }
                    }
                    2 -> {
                        if (AppCompatDelegate.getDefaultNightMode() != AppCompatDelegate.MODE_NIGHT_YES) {
                            settingsPrefs.edit().putInt(PREFERENCE_THEME_KEY, i)
                                .apply()
                            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
                        }
                    }
                }
                currentTheme = i
                setCurrentThemeText()
                dialogInterface.dismiss()
            }
        }.show()
    }

    private fun privacyPolicyOnClick() =
        binding.clSettingsPrivacyPolicy.setOnClickListener {
            Intent(Intent.ACTION_VIEW, Uri.parse(URL_PRIVACY_POLICY)).apply {
                this.resolveActivity(requireContext().packageManager)?.let {
                    startActivity(this)
                } ?: Snackbar.make(
                    binding.root,
                    requireContext().getString(R.string.intent_error_app_not_found),
                    LENGTH_LONG
                ).show()
            }
        }

    private fun logoutOnClick() = binding.btnLogout.setOnClickListener {
        logout()
    }

    private fun logout() = CoroutineScope(Dispatchers.IO).launch {
        try {
            firebaseAuth.signOut()
            Log.i("logout", "${currentUser?.email} successfully logged out.")

            requireContext().getSharedPreferences(
                PREFERENCE_LOGIN, Context.MODE_PRIVATE
            ).edit().apply {
                remove(PREFERENCE_IS_LOGGED_IN_KEY)
                apply()
            }

            startActivity(Intent(requireContext(), LandingActivity::class.java))
            requireActivity().finish()
        } catch (e: Exception) {
            Log.e("logout", "Exception: ${e.message}")
        }
    }

    private fun showLoadingAnimation() {
        binding.btnAccountVerification.visibility = GONE
        binding.cpiVerify.visibility = VISIBLE

    }

    private fun hideLoadingAnimation() {
        binding.cpiVerify.visibility = GONE
        binding.btnAccountVerification.visibility = VISIBLE
    }

    private fun adInit() = TapsellPlus.initialize(requireContext(), TAPSELL_KEY,
        object : TapsellPlusInitListener {
            override fun onInitializeSuccess(adNetworks: AdNetworks?) {
                Log.i("adInit", "onInitializeSuccess: ${adNetworks?.name}")
                val adHolder = TapsellPlus.createAdHolder(
                    activity, binding.flAdContainer, R.layout.ad_banner_settings
                )
                requestAdCounter = 0
                adHolder?.let { requestNativeAd(it) }
            }

            override fun onInitializeFailed(
                adNetworks: AdNetworks?, adNetworkError: AdNetworkError?
            ) {
                Log.e(
                    "adInit",
                    "onInitializeFailed: ${adNetworks?.name}, error: ${adNetworkError?.errorMessage}"
                )
            }
        })

    private fun requestNativeAd(adHolder: AdHolder) {
        TapsellPlus.requestNativeAd(requireActivity(),
            SETTINGS_NATIVE_ZONE_ID, object : AdRequestCallback() {
                override fun response(tapsellPlusAdModel: TapsellPlusAdModel?) {
                    super.response(tapsellPlusAdModel)
                    Log.i("requestNativeAd", "response: ${tapsellPlusAdModel.toString()}")
                    _binding?.let {
                        tapsellPlusAdModel?.let {
                            responseId = it.responseId
                            showNativeAd(adHolder, responseId!!)
                        }
                    }
                }

                override fun error(error: String?) {
                    super.error(error)
                    Log.e("requestNativeAd", "Error: $error")
                    if (requestAdCounter < 3) {
                        requestAdCounter++
                        requestNativeAd(adHolder)
                    }
                }
            })
    }

    private fun showNativeAd(adHolder: AdHolder, responseId: String) {
        binding.groupAd.visibility = VISIBLE
        TapsellPlus.showNativeAd(requireActivity(),
            responseId, adHolder, object : AdShowListener() {
                override fun onOpened(tapsellPlusAdModel: TapsellPlusAdModel?) {
                    super.onOpened(tapsellPlusAdModel)
                }

                override fun onClosed(tapsellPlusAdModel: TapsellPlusAdModel?) {
                    super.onClosed(tapsellPlusAdModel)
                }
            })
    }

    private fun destroyAd() = responseId?.let {
        TapsellPlus.destroyNativeBanner(requireActivity(), it)
    }
}